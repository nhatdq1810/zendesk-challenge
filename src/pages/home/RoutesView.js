import React from 'react';
import PropTypes from 'prop-types';
import { Timeline, Collapse, Icon } from 'antd';

const getDirection = (line, fromStation, toStation) => (
  <b>
    Take line {line}
    &nbsp;from station {fromStation} to station {toStation}
  </b>
);

export default function RoutesView({
  orderedRoutes, orderedRoutesByLines, originStation, destStation,
}) {
  return (
    <Collapse defaultActiveKey="0" bordered={false}>
      {orderedRoutes.map((route, i) => {
        const displayedRoute = [];
        let routeIntro = '';

        if (route.length === 1) {
          if (route[0].name === destStation) {
            routeIntro = `Via line ${orderedRoutesByLines[i][0]}, station ${destStation}`;
            displayedRoute.push(
              <Timeline.Item
                color="green"
                dot={<Icon type="logout" />}
                key={`${originStation} - ${destStation}`}
              >
                {getDirection(orderedRoutesByLines[i][0], originStation, destStation)}
              </Timeline.Item>,
            );
          } else {
            routeIntro = `Via line ${orderedRoutesByLines[i][0]}, station ${route[0].name}`;
            displayedRoute.push(
              <Timeline.Item
                color="red"
                dot={<Icon type="login" />}
                key={`${originStation} - ${route[0].name}`}
              >
                {getDirection(orderedRoutesByLines[i][0], originStation, route[0].name)}
              </Timeline.Item>,
              <Timeline.Item
                color="green"
                dot={<Icon type="logout" />}
                key={`${route[0].name} - ${destStation}`}
              >
                {getDirection(orderedRoutesByLines[i][1], route[0].name, destStation)}
              </Timeline.Item>,
            );
          }
        } else {
          route.forEach((station, j) => {
            if (j === 0) {
              routeIntro = `Via line ${orderedRoutesByLines[i][j]}, station ${station.name}`;
              displayedRoute.push(
                <Timeline.Item
                  color="red"
                  dot={<Icon type="login" />}
                  key={`${originStation} - ${station.name}`}
                >
                  {getDirection(orderedRoutesByLines[i][j], originStation, station.name)}
                </Timeline.Item>,
              );
            } else {
              displayedRoute.push(
                <Timeline.Item key={`${route[j - 1].name} - ${station.name}`}>
                  {getDirection(orderedRoutesByLines[i][j], route[j - 1].name, station.name)}
                </Timeline.Item>,
              );
            }

            if (j === route.length - 1) {
              displayedRoute.push(
                <Timeline.Item
                  color="green"
                  dot={<Icon type="logout" />}
                  key={`${station.name} - ${destStation}`}
                >
                  {getDirection(orderedRoutesByLines[i][j + 1], station.name, destStation)}
                </Timeline.Item>,
              );
            }
          });
        }

        return (
          <Collapse.Panel key={`${routeIntro}-${i}`} header={routeIntro}>
            <Timeline>
              {displayedRoute}
            </Timeline>
          </Collapse.Panel>
        );
      })}
    </Collapse>
  );
}

RoutesView.propTypes = {
  orderedRoutes: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.object)).isRequired,
  orderedRoutesByLines: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.string)).isRequired,
  originStation: PropTypes.string.isRequired,
  destStation: PropTypes.string.isRequired,
};
