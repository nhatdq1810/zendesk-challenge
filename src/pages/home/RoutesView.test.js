import React from 'react';
import renderer from 'react-test-renderer';
import RoutesView from './RoutesView';

describe('RoutesView', () => {
  it('should render one route contains only destination station', () => {
    const snapshot = renderer.create(<RoutesView
      orderedRoutes={[
        [{ name: 'final', CC: 2, DT: 1 }],
      ]}
      orderedRoutesByLines={[
        ['CC', 'DT'],
      ]}
      originStation="begin"
      destStation="final"
    />);

    expect(snapshot).toMatchSnapshot();
  });

  it('should render one route contains only one station', () => {
    const snapshot = renderer.create(<RoutesView
      orderedRoutes={[
        [{ name: 'A', CC: 2, DT: 1 }],
      ]}
      orderedRoutesByLines={[
        ['CC', 'DT'],
      ]}
      originStation="begin"
      destStation="final"
    />);

    expect(snapshot).toMatchSnapshot();
  });

  it('should render one route contains two interchange stations', () => {
    const snapshot = renderer.create(<RoutesView
      orderedRoutes={[
        [{ name: 'A', CC: 2, CE: 1 }, { name: 'B', CE: 4, DT: 2 }],
      ]}
      orderedRoutesByLines={[
        ['CC', 'CE', 'DT'],
      ]}
      originStation="begin"
      destStation="final"
    />);

    expect(snapshot).toMatchSnapshot();
  });
});
