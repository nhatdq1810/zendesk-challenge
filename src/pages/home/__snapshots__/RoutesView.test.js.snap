// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoutesView should render one route contains only destination station 1`] = `
<div
  className="ant-collapse ant-collapse-borderless"
  role={null}
>
  <div
    className="ant-collapse-item"
  >
    <div
      aria-expanded="false"
      className="ant-collapse-header"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <i
        className="anticon anticon-right arrow"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="right"
          fill="currentColor"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
          />
        </svg>
      </i>
      Via line CC, station final
    </div>
  </div>
</div>
`;

exports[`RoutesView should render one route contains only one station 1`] = `
<div
  className="ant-collapse ant-collapse-borderless"
  role={null}
>
  <div
    className="ant-collapse-item"
  >
    <div
      aria-expanded="false"
      className="ant-collapse-header"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <i
        className="anticon anticon-right arrow"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="right"
          fill="currentColor"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
          />
        </svg>
      </i>
      Via line CC, station A
    </div>
  </div>
</div>
`;

exports[`RoutesView should render one route contains two interchange stations 1`] = `
<div
  className="ant-collapse ant-collapse-borderless"
  role={null}
>
  <div
    className="ant-collapse-item"
  >
    <div
      aria-expanded="false"
      className="ant-collapse-header"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <i
        className="anticon anticon-right arrow"
      >
        <svg
          aria-hidden="true"
          className=""
          data-icon="right"
          fill="currentColor"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
          />
        </svg>
      </i>
      Via line CC, station A
    </div>
  </div>
</div>
`;
